noxfile.py

```diff
@@ -0,0 +1,279 @@
+import os
+import platform
+import shutil
+import nox
+
+GCC_VERSIONS = ["gcc-5", "gcc-6", "gcc-8", "clang-10"]
+GCC_VERSION2USE = os.environ.get("CC", "gcc-5")
+DEFAULT_TEST_DIRECTORIES = ["doc", "gcovr"]
+BLACK_CONFORM_FILES = [
+    "noxfile.py",
+    "gcovr/gcov.py",
+    "gcovr/gcov_parser.py",
+]
+
+nox.options.sessions = [
+    "qa",
+    "lint",
+    "doc",
+    "tests_version({})".format(GCC_VERSION2USE),
+]
+
+
+def set_environment(session: "nox.session", cc: str, check: bool = True) -> None:
+    if check and (shutil.which(cc) is None):
+        session.env["CC_REFERENCE"] = cc
+        cc = "gcc"
+        cxx = "g++"
+        gcov = "gcov"
+    else:
+        cxx = cc.replace("clang", "clang++").replace("gcc", "g++")
+        if cc.startswith("clang"):
+            gcov = cc.replace("clang", "llvm-cov") + " gcov"
+        else:
+            gcov = cc.replace("gcc", "gcov")
+    session.env["GCOVR_TEST_SUITE"] = "1"
+    session.env["CC"] = cc
+    session.env["CFLAGS"] = "--this_flag_does_not_exist"
+    session.env["CXX"] = cxx
+    session.env["CXXFLAGS"] = "--this_flag_does_not_exist"
+    session.env["GCOV"] = gcov
+
+
+@nox.session
+def qa(session: "nox.session") -> None:
+    for session_id in nox.options.sessions:
+        session.log("Notify session {}".format(session_id))
+        session.notify(session_id)
+
+
+@nox.session
+def lint(session: "nox.session") -> None:
+    session.install("flake8")
+    if platform.python_implementation() == "CPython":
+        session.install("black")
+    if session.posargs:
+        args = session.posargs
+    else:
+        args = DEFAULT_TEST_DIRECTORIES
+    session.run("flake8", *args)
+
+    if platform.python_implementation() == "CPython":
+        if session.posargs:
+            session.run("python", "-m", "black", *session.posargs)
```

Comment:
I would not expect `nox -s lint -- doc` to apply any changes. I would expect that to restrict the linting checks to the doc directory. If I wanted to reformat the files in the doc directory, I'd run `nox -s black -- doc`.

Reply:
I'll move the default args to the black session (needed for the GitHub Actions) and for the lint I'll notify the black session with the needed arguments.

Reply:
In the Makefile lint only called flake8 and check-format was black. If lint also calls black we don't need the check-format step in the CI. What do you think?

Reply:
I think the idea of keeping these steps separate in the CI was to get better error messages about which step of the QA process failed. Right now, running Black isn't necessarily attractive because it produces huge diffs that fill up the terminal. But long-term, it makes sense to just run all linting tools together as a single Nox session so that developers don't have to manually select a particular step. Not sure what the correct tradeoffs are here, so for the time being I'd just leave everything as it is now.